### ISSUE TITLE
a question regarding right usage of  Aws::Auth::AWSCredentialsProvider interface

### ISSUE THREAD

**Issue by Eliyahu-Machluf** [aws/aws-sdk-cpp#902]
Hi,

I'm trying to use the Aws::Auth::AWSCredentialsProvider interface. I've created a new class inheriting from this class. However my method GetAWSCredentials can fail from time to time (as I'm talking with Active directory, get SAML, and get STS credentials which are valid for 1 hour only).
I don't know what shall I return when failed to retrieve the credentials. i.e. what shall I return from GetAWSCredentials on failure.
I thought the method GetAWSCredentials shall throw an exception to indicate an error occurred. However the using code at auth::AWSAuthSigner.cpp does not seem to handle any such exception.
when I do action which run on same thread like list bucket, get object properties etc. I can wrap the calling code with try and catch, and this catch the exception. However when I use the transfer manager, it uses its own internal threads for file upload. and if GetAWSCredentials fails for these internal threads, no one catch it and this cause the process to crash.

To summarize, my question is what shall I return from GetAWSCredentials on error, and note that for transfer manager upload file scenario, there are internal threads which are not managed by me.

Thanks.

Eliyahu

**Comment by marcomagdy** [comment]
Yeah, that's a wrinkle in the interface. It needs to have a way to signal errors. We don't use exceptions in the SDK, so we should have used the Outcome pattern.

As a workaround, you can define a static AwsCredentials object as a sentinel and return that in the case of errors. In your calling code, you would check if the returned object is the sentinel, either by equality comparison or simply checking its memory address.

Does that work for you?

**Comment by marcomagdy** [comment]
You’re welcome.
